### PR TITLE
Start documenting public parts of all code

### PR DESCRIPTION
**Pull request by badboy**
This is nowhere near complete, but it gets us started.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

File: glean-core/src/metrics/counter.rs [hunk 1]
```diff
@@ -25,10 +28,14 @@ impl MetricType for CounterMetric {
 }
 
 impl CounterMetric {
+    /// Create a new metric
     pub fn new(meta: CommonMetricData) -> Self {
         Self { meta }
     }
 
+    /// Increase the counter by `amount`
+    ///
```
Comment (Dexterp37) [inline glean-core/src/metrics/counter.rs]: Can we document arguments here, using the format we're using in other parts of the code?

File: glean-core/src/metrics/mod.rs [hunk 2]
```diff
@@ -2,6 +2,10 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
+//! # Metrics
+//!
+//! Glean supports different metric types to store data.
```
Comment (Dexterp37) [inline glean-core/src/metrics/mod.rs]: nit: should this be more like "The different metric types supported by the Glean SDK to handle data"?

File: glean-core/src/metrics/mod.rs [hunk 3]
```diff
@@ -30,27 +34,44 @@ pub use self::string_list::StringListMetric;
 pub use self::time_unit::TimeUnit;
 pub use self::uuid::UuidMetric;
 
+/// The available metrics
+///
+/// This is the in-memory and persisted layout of a metric
 #[derive(Serialize, Deserialize, Clone, Debug)]
 pub enum Metric {
+    /// A boolean metric. See [`BooleanMetric`](struct.BooleanMetric.html) for more information.
```
Comment (Dexterp37) [inline glean-core/src/metrics/mod.rs]: Ah, doesn't autolink/discover things?

File: glean-core/src/metrics/mod.rs [hunk 4]
```diff
@@ -30,27 +34,44 @@ pub use self::string_list::StringListMetric;
...
+    ///
+    /// This depends on the metrics own state, as determined by its metadata,
+    /// and whether upload is enabled on the Glean object.
     fn should_record(&self, glean: &Glean) -> bool {
         glean.is_upload_enabled() && self.meta().should_record()
     }
 }
 
 impl Metric {
+    /// The category the metric fits into
```
Comment (Dexterp37) [inline glean-core/src/metrics/mod.rs]: Can you expand on what is this used for? (i.e. to decide where to put stuff in the ping?)

File: glean-core/src/metrics/string.rs [hunk 5]
```diff
@@ -11,6 +11,9 @@ use crate::Glean;
 
 const MAX_LENGTH_VALUE: usize = 100;
 
+/// # A string metric
+///
+/// Record an Unicode string value with arbitrary content.
```
Comment (Dexterp37) [inline glean-core/src/metrics/string.rs]: Should we mention that it's length limited and point to the `MAX_LENGTH_VALUE` above?

File: glean-core/src/metrics/string.rs [hunk 6]
```diff
@@ -27,10 +30,12 @@ impl MetricType for StringMetric {
 }
 
 impl StringMetric {
+    /// Create a new metric
     pub fn new(meta: CommonMetricData) -> Self {
         Self { meta }
     }
 
+    /// Set to the specified value
```
Comment (Dexterp37) [inline glean-core/src/metrics/string.rs]: nit: document the parameters

File: glean-core/src/metrics/boolean.rs [hunk 7]
```diff
@@ -24,10 +27,12 @@ impl MetricType for BooleanMetric {
 }
 
 impl BooleanMetric {
+    /// Create a new metric
```
[the remark this answers is not in the document]
Comment (badboy) [reply glean-core/src/metrics/boolean.rs]: That is an excellent question. Let me see what the standard library does.

File: glean-core/src/metrics/mod.rs (hunk 3 above)
Comment (badboy) [reply glean-core/src/metrics/mod.rs]: So there is an [intra-doc links RFC](https://github.com/rust-lang/rfcs/blob/master/text/1946-intra-rustdoc-links.md) and it seems part of this is implemented, but I don't think it's stable yet (at least I didn't get it to work, I'll ask around)

Comment (Dexterp37) [reply glean-core/src/metrics/mod.rs]: Ah, interesting. I'm fine with leaving this as it is for now, as long as we file a bug to fix it in the future :)

File: glean-core/src/metrics/boolean.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (badboy) [reply glean-core/src/metrics/boolean.rs]: Let's require it (e.g. [HashMap does it](https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#684))

[the remark this answers is not in the document]
Comment (Dexterp37) [reply glean-core/src/metrics/boolean.rs]: Thank you for checking this out!

**Pull request merged**